### DISCUSSION TITLE
How to store value data in a sample2D texture

### DISCUSSION THREAD

**Discussion by fanbinnet** [google/filament#4996] [Q&A]
I'm writing a GPU particles demo base on filament.
The demo need store particle pos and vel in a texture.
Can anyone give me some suggestion for saving data to texture ?
Thanks very much!

**Comment by fanbinnet** [comment]
I use below code to init the position，but the value in shader is not
right，     I have used the RGBA32F format
  ```
val buffer = ByteBuffer.allocateDirect(AMOUNT * 4 * FLOAT_SIZE)
        for (i in 0 until AMOUNT) {
            val radius = (0.5 + Math.random() * 0.5) * 50
            val phi = (Math.random() - 0.5) * Math.PI
            val theta = Math.random() * Math.PI * 2
            val red = (radius * Math.cos(theta) * Math.cos(phi)).toFloat()
            val green = (radius * Math.sin(phi)).toFloat()
            val blue = (radius * Math.sin(theta) * Math.cos(phi)).toFloat()
            val alpha = (Math.random()).toFloat()
            buffer.putFloat(red)
            buffer.putFloat(green)
            buffer.putFloat(blue)
            buffer.putFloat(alpha)
        }
        buffer.flip()
        val texture = Texture.Builder()
            .width(FBO_WIDTH)
            .height(FBO_HEIGHT)
            .sampler(Texture.Sampler.SAMPLER_2D)
            .format(Texture.InternalFormat.RGBA32F)
            .levels(1)
            .build(engine)

        val descriptor = Texture.PixelBufferDescriptor(
            buffer,
            Texture.Format.RGBA,
            Texture.Type.FLOAT
        )

        texture.setImage(engine, 0, descriptor)
```

**Reply by romainguy** [reply]
How do you sample it? RGBA32F is not filterable without extensions on GLES.

**Reply by fanbinnet** [reply]
If can I use the RGBA16F ？Is the RGBA16F filterable ?

**Reply by bejado** [reply]
`RGBA16F` is [filterable](https://www.khronos.org/registry/OpenGL-Refpages/es3.0/html/glTexImage2D.xhtml), but seems like you *wouldn't* want filtering if you're using the texture to store data. Have you tried using [`texelFetch`](https://www.khronos.org/registry/OpenGL-Refpages/es3.0/html/texelFetch.xhtml) to access the texture? What values are you seeing in the shader?
